artdaq/ArtModules/ArtdaqSharedMemoryService_service.cc

```diff
@@ -181,70 +175,135 @@ std::unordered_map<artdaq::Fragment::type_t, std::unique_ptr<artdaq::Fragments>>
+			if (!output_event->fragments[type])
 			{
 				TLOG(TLVL_WARNING) << "Error retrieving Fragments from shared memory! (Most likely due to a buffer overwrite) Retrying...";
 				incoming_events_->ReleaseBuffer();
-				recvd_fragments.clear();
+				output_event->fragments.clear();
 				continue;
 			}
-			/* Events coming out of the EventStore are not sorted but need to be
-	   sorted by sequence ID before they can be passed to art.
-	*/
-			std::sort(recvd_fragments[type]->begin(), recvd_fragments[type]->end(), artdaq::fragmentSequenceIDCompare);
+			// Events coming out of the EventStore are not sorted but need to be sorted by sequence ID before they can be passed to art.
+			std::sort(output_event->fragments[type]->begin(), output_event->fragments[type]->end(), artdaq::fragmentSequenceIDCompare);
 		}
-		TLOG(TLVL_DEBUG + 33) << "ReceiveEvent: Releasing buffer";
+		TLOG(TLVL_DEBUG + 33) << "ReadEventFromSharedMemory: Releasing buffer";
 		incoming_events_->ReleaseBuffer();
 	}
 
+	TLOG(TLVL_DEBUG + 33) << "ReadEventFromSharedMemory END";
+	return output_event;
+}
+
+std::shared_ptr<ArtdaqEvent> ArtdaqSharedMemoryService::ReceiveEvent(bool broadcast)
+{
+	TLOG(TLVL_DEBUG + 33) << "ReceiveEvent BEGIN";
+	std::shared_ptr<ArtdaqEvent> output_event;
+
+	while (output_event == nullptr)
+	{
+		// If we experienced a timeout, drain any held Start/End Run/SubRun events
+		if (last_read_timeout_)
+		{
+			if (event_ordering_.size() > 0)
+			{
+				output_event = event_ordering_.front();
+				event_ordering_.pop_front();
+				break; // while(output_event == nullptr)
+			}
+		}
+
+		bool eventInEventOrder = false;
+		for (auto it = event_ordering_.begin(); it != event_ordering_.end(); ++it)
+		{
+			auto type = (*it)->FirstFragmentType();
+
+            // We will automatically start draining if we have an Init, EndOfData, or Event message
+			if (type == artdaq::Fragment::StartOfRunFragmentType || type == artdaq::Fragment::StartOfSubrunFragmentType 
+                || type == artdaq::Fragment::EndOfSubrunFragmentType || type == artdaq::Fragment::EndOfRunFragmentType
+                || type == artdaq::Fragment::RunDataFragmentType || type == artdaq::Fragment::SubrunDataFragmentType)
+			{
+				continue; // for (auto it = event_ordering_.begin(); it != event_ordering_.end(); ++it)
+			}
+			eventInEventOrder = true;
+			break; // for (auto it = event_ordering_.begin(); it != event_ordering_.end(); ++it)
+		}
+		if (eventInEventOrder)
+		{
+			output_event = event_ordering_.front();
+			event_ordering_.pop_front();
+			break; // while(output_event == nullptr)
+        }
+
+		auto next_event = ReadEventFromSharedMemory(broadcast);
+		if (next_event == nullptr)
+		{
+			if (event_ordering_.size() > 0)
+			{
+				output_event = event_ordering_.front();
+				event_ordering_.pop_front();
+			}
+			else
+			{
+                // Will return nullptr
+				break; // while(output_event == nullptr)
```

Comment:
I like these comments indicating on which scope the break/continue is operating.